src/EllCrv/EllCrv.jl

```diff
@@ -707,124 +986,785 @@ function *(n::Int, P::EllCrvPt)
+    m = div(n-1,2)
+    part1 = divpol_g_short(E,m+2,x)  * divpol_g_short(E,m,x)^3
+    part2 = divpol_g_short(E,m-1,x) * divpol_g_short(E,m+1,x)^3
+    if mod(m,2) == 0
+      return B6sqr * part1 - part2
+    else
+      return part1 - B6sqr * part2
+    end
   end
 end
 
-function replace_all_squares(f, g)
-    # assumes that f is in Z[x,y^2] and g in Z[x]. Replaces y^2 with g.
-    # the result will be in Z[x]
-    z = zero(parent(g)) # this is the zero in Z[x]
-    d = div(degree(f), 2) # degree of f in y^2 should be even
-    for i in 0:d
-        z = z + coeff(f, 2*i)*g^i
+function divpol_g(E::EllCrv, n::S, x = PolynomialRing(base_field(E),"x")[2]) where S<:Union{Integer, fmpz}
+  
+  Kx = parent(x)
+  
+  b2, b4, b6, b8 = E.b_invars
+  B4 = 6*x^2+b2*x+b4
+  B6sqr = (4*x^3+b2*x^2+2*b4*x+b6)^2
+  B8 = 3*x^4 + b2*x^3 + 3*b4*x^2 + 3*b6*x + b8
+
+  
+  if n == 1 || n ==2
+    return one(Kx)
+  elseif n == 3
+    return B8
+  elseif n == 4
+    return -B6sqr+B4*B8
+  elseif mod(n,2) == 0
+    m = div(n-2,2)
+    return divpol_g(E,m+3,x)*divpol_g(E,m+1,x)*divpol_g(E,m,x)^2 - divpol_g(E,m+1,x)*divpol_g(E,m-1,x)*divpol_g(E,m+2,x)^2
+  else
+    m = div(n-1,2)
+    part1 = divpol_g(E,m+2,x)  * divpol_g(E,m,x)^3
+    part2 = divpol_g(E,m-1,x) * divpol_g(E,m+1,x)^3
+    if mod(m,2) == 0
+      return B6sqr * part1 - part2
+    else
+      return part1 - B6sqr * part2
     end
-    return z
+  end
 end
 
-function replace_all_squares_modulo(f, g, F)
-    # assumes that f is in Z[x,y^2] and g in Z[x]. Replaces y^2 with g.
-    # the result will be in Z[x]
-    z = zero(parent(g)) # this is the zero in Z[x]
-    d = div(degree(f), 2) # degree of f in y^2 should be even
-    for i in 0:d
-        z = z + coeff(f, 2*i)*powermod(g, i, F)
+
+################################################################################
+#
+#  Multiplication (as maps) and division by m 
+#
+################################################################################
+
+function multiplication_by_m_numerator(E::EllCrv, m::S, x = PolynomialRing(base_field(E),"x")[2]) where S<:Union{Integer, fmpz}
+
+  Kx = parent(x)
+  
+  b2, b4, b6, b8 = b_invars(E)
+  B6= 4*x^3+b2*x^2+2*b4*x+b6
+  
+  psi_mmin = division_polynomial_univariate(E, m-1, x)[2]
+  psi_m = division_polynomial_univariate(E, m, x)[2]
+  psi_mplus = division_polynomial_univariate(E, m+1, x)[2]
+  
+   if mod(m,2) == 0
+      return x * B6 * psi_m^2 - psi_mmin * psi_mplus
+    else
+      return x * psi_m^2-B6 * psi_mmin * psi_mplus
     end
-    return z
 end
 
-# here is an example: f = _Zxy(_x)*y^2 + _Zxy(_x + 1)*y^4; EC.replace_all_squares(f, x^3)
-# this should give _x^7+_x^6+_x^4
+function multiplication_by_m_denominator(E::EllCrv, m::S, x = PolynomialRing(base_field(E),"x")[2]) where S<:Union{Integer, fmpz}
+  
+  Kx = parent(x)
+  
+  b2, b4, b6, b8 = b_invars(E)
+  B6= 4*x^3+b2*x^2+2*b4*x+b6
+  
+  psi_m = division_polynomial_univariate(E, m, x)[2]
+
+   if mod(m,2) == 0
```

Comment:
I don't mind too much how things are indented, but I think it'd be nice if it was at least consistent within functions... Hecke seems to use 2 space indentation, so perhaps use that here, too?